Azure/sonic-utilities
# [warm-reboot] Fix failures of warm reboot on disconnect of ssh session

* Starting the script in background mode and detaching the process
  from terminal session to prevent failures, caused by closing or
  sudden disconnecting of terminal session.
* Redirecting output of the script to the file to prevent failures,
  when the script tries to write an output to file descriptor of
  the nonexistent terminal session.
* Adding new parameter to script to be able to explicitly run the
  script in foreground mode with output to the terminal.
* Updating the command reference doc according to the changes.

Signed-off-by: Maksym Belei <Maksym_Belei@jabil.com>



#### What I did
Resolves https://github.com/Azure/sonic-buildimage/issues/7127
Fixed failures of warm reboot, when the SSH session is being disconnected.
As the script will now be executed in background mode by default, added parameter to explicitly run it as usual, in foreground mode.
Updated command reference according to the changes.

#### How I did it
By restarting the script in background mode with detaching it from the terminal session. 
All the output has redirected to file /var/log/warm-reboot.txt for warm-reboot case, or /var/log/fast-reboot.txt for fast-reboot, depends on  `REBOOT_TYPE`. This will prevent crashes of the script in case, when it will try to write some data to the file descriptor of the disconnected terminal session.

#### How to verify it
1. Connect to the switch with SSH;
2. Execute `sudo warm-reboot -v`;
3. See the current progress of warm reboot with `cat /var/log/warm-reboot.txt`;
4. Close SSH connection before warm reboot finish;
Warm reboot should finish successfully, in spite of status of the SSH session.

#### New command output (if the output of a command-line utility has changed)
The script will be running in background detached mode with output to the file. The related log will be shown in terminal before restarting in background mode:
```
admin@sonic:~$ sudo warm-reboot
Detaching the process from the terminal session. Redirecting output to /var/log/warm-reboot.txt.
```
All the usual logs will be written to warm-reboot.txt.
